src/node_webstorage.cc

```diff
@@ -0,0 +1,654 @@
+  }
+
+  r = sqlite3_exec(_db, sql, 0, 0, nullptr);
+  if (r != SQLITE_OK) {
+    THROW_SQLITE_ERROR(env(), r);
+    return false;
+  }
+
+  return true;
+}
+
+void Storage::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Realm* realm = Realm::GetCurrent(args);
+
+  if (!args[0]->StrictEquals(realm->isolate_data()->constructor_key_symbol())) {
+    THROW_ERR_ILLEGAL_CONSTRUCTOR(env);
+    return;
+  }
+
+  CHECK(args.IsConstructCall());
+  CHECK(args[1]->IsString());
+  new Storage(env, args.This(), args[1].As<String>());
+}
+
+void Storage::Clear() {
+  if (!Open()) {
+    return;
+  }
+
+  static const char sql[] = "DELETE FROM nodejs_webstorage";
+  sqlite3_stmt* stmt = nullptr;
+  auto cleanup = OnScopeLeave([&]() { sqlite3_finalize(stmt); });
+  CHECK_ERROR_OR_THROW(
+      env(), sqlite3_prepare_v2(_db, sql, -1, &stmt, 0), SQLITE_OK, void());
+  CHECK_ERROR_OR_THROW(env(), sqlite3_step(stmt), SQLITE_DONE, void());
+}
+
+Local<Array> Storage::Enumerate() {
+  if (!Open()) {
+    return Local<Array>();
+  }
+
+  static const char sql[] = "SELECT key FROM nodejs_webstorage";
+  sqlite3_stmt* stmt = nullptr;
+  auto cleanup = OnScopeLeave([&]() { sqlite3_finalize(stmt); });
+  int r = sqlite3_prepare_v2(_db, sql, -1, &stmt, 0);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, Local<Array>());
+  std::vector<Local<Value>> values;
+  while ((r = sqlite3_step(stmt)) == SQLITE_ROW) {
+    CHECK(sqlite3_column_type(stmt, 0) == SQLITE_BLOB);
+    values.emplace_back(
+        String::NewFromUtf8(
+            env()->isolate(),
+            reinterpret_cast<const char*>(sqlite3_column_blob(stmt, 0)),
+            v8::NewStringType::kNormal,
+            static_cast<int>(sqlite3_column_bytes(stmt, 0))).ToLocalChecked());
+  }
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_DONE, Local<Array>());
+  return Array::New(env()->isolate(), values.data(), values.size());
+}
+
+Local<Value> Storage::Length() {
+  if (!Open()) {
+    return Local<Value>();
+  }
+
+  static const char sql[] = "SELECT count(*) FROM nodejs_webstorage";
+  sqlite3_stmt* stmt = nullptr;
+  auto cleanup = OnScopeLeave([&]() { sqlite3_finalize(stmt); });
+  int r = sqlite3_prepare_v2(_db, sql, -1, &stmt, 0);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, Local<Value>());
+  CHECK_ERROR_OR_THROW(env(), sqlite3_step(stmt), SQLITE_ROW, Local<Value>());
+  CHECK(sqlite3_column_type(stmt, 0) == SQLITE_INTEGER);
+  int result = sqlite3_column_int(stmt, 0);
+  return Integer::New(env()->isolate(), result);
+}
+
+Local<Value> Storage::Load(Local<Name> key) {
+  if (key->IsSymbol()) {
+    auto symbol_map = _symbols.Get(env()->isolate());
+    MaybeLocal<Value> result = symbol_map->Get(env()->context(), key);
+    return result.FromMaybe(Local<Value>());
+  }
+
+  if (!Open()) {
+    return Local<Value>();
+  }
+
+  static const char sql[] =
+      "SELECT value FROM nodejs_webstorage WHERE key = ? LIMIT 1";
+  sqlite3_stmt* stmt = nullptr;
```

Comment:
Eventually I wonder if it would make sense to add some basic caching here for these statements? It'll be something worth doing some perf profiling on.